tortoise/verifying_tortoise.go

```diff
@@ -49,73 +101,132 @@ type turtle struct {
 
 	Verified types.LayerID
 
-	// use 2D array to be able to iterate from latest elements easily
-	BlockOpinionsByLayer map[types.LayerID]map[types.BlockID]Opinion // records hdist, for each block, its votes about every
-	// previous block
+	// this matrix stores the opinion of each block about other blocks. blocks are indexed by layer.
+	// it stores good and bad blocks
+	BlockOpinionsByLayer map[types.LayerID]map[types.BlockID]Opinion
 
-	// TODO: Tal says: We keep a vector containing our vote totals (positive and negative) for every previous block
-	// that's not needed here, probably for self healing?
+	// how often we want to rerun from genesis
+	RerunInterval time.Duration
 }
 
 // SetLogger sets the Log instance for this turtle
-func (t *turtle) SetLogger(log2 log.Log) {
-	t.logger = log2
+func (t *turtle) SetLogger(logger log.Log) {
+	t.logger = logger
 }
 
-// newTurtle creates a new verifying tortoise algorithm instance. XXX: maybe rename?
-func newTurtle(bdp blockDataProvider, hdist uint32, avgLayerSize int) *turtle {
-	t := &turtle{
-		logger:               log.NewNop(),
+// newTurtle creates a new verifying tortoise algorithm instance
+func newTurtle(
+	bdp blockDataProvider,
+	atxdb atxDataProvider,
+	clock layerClock,
+	hdist,
+	zdist,
+	confidenceParam,
+	windowSize uint32,
+	avgLayerSize int,
+	globalThreshold,
+	localThreshold uint8,
+	rerun time.Duration,
+) *turtle {
+	return &turtle{
+		logger:               log.NewDefault("trtl"),
 		Hdist:                hdist,
+		Zdist:                zdist,
+		ConfidenceParam:      confidenceParam,
+		WindowSize:           windowSize,
+		GlobalThreshold:      globalThreshold,
+		LocalThreshold:       localThreshold,
 		bdp:                  bdp,
+		atxdb:                atxdb,
+		clock:                clock,
+		Last:                 types.NewLayerID(0),
 		AvgLayerSize:         avgLayerSize,
 		GoodBlocksIndex:      make(map[types.BlockID]struct{}),
 		BlockOpinionsByLayer: make(map[types.LayerID]map[types.BlockID]Opinion, hdist),
 		MaxExceptions:        int(hdist) * avgLayerSize * 100,
+		RerunInterval:        rerun,
 	}
-	return t
 }
 
-func (t *turtle) init(genesisLayer *types.Layer) {
+// cloneTurtle creates a new verifying tortoise instance using the params of this instance
+func (t *turtle) cloneTurtle() *turtle {
```

Comment:
the name got me confused at first, thought you were cloning the tables as well, maybe mention you're cloning or copying configuration in the name.